src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedList`1.cs

```diff
@@ -0,0 +1,794 @@
+    /// <remarks>
+    /// This class implement a list which is allocated in segments, to avoid large lists to go into LOH.
+    /// </remarks>
+    internal sealed class SegmentedList<T> : ICollection<T>, IReadOnlyList<T>
+    {
+        private readonly int _segmentSize;
+        private readonly int _segmentShift;
+        private readonly int _offsetMask;
+
+        private int _capacity;
+        private int _count;
+        private T[][] _items;
+
+        /// <summary>
+        /// Constructs SegmentedList.
+        /// </summary>
+        /// <param name="segmentSize">Segment size</param>
+        public SegmentedList(int segmentSize)
+            : this(segmentSize, 0)
+        {
+        }
+
+        /// <summary>
+        /// Constructs SegmentedList.
+        /// </summary>
+        /// <param name="segmentSize">Segment size</param>
+        /// <param name="initialCapacity">Initial capacity</param>
+        public SegmentedList(int segmentSize, int initialCapacity)
+        {
+            if (segmentSize <= 1 || (segmentSize & (segmentSize - 1)) != 0)
+            {
+                throw new ArgumentOutOfRangeException("segment size must be power of 2 greater than 1");
+            }
+
+            _segmentSize = segmentSize;
+            _offsetMask = segmentSize - 1;
+            _segmentShift = 0;
+
+            while (0 != (segmentSize >>= 1))
+            {
+                _segmentShift++;
+            }
+
+            if (initialCapacity > 0)
+            {
+                initialCapacity = _segmentSize * ((initialCapacity + _segmentSize - 1) / _segmentSize);
+                _items = new T[initialCapacity >> _segmentShift][];
+                for (var i = 0; i < _items.Length; i++)
+                {
+                    _items[i] = new T[_segmentSize];
+                }
+
+                _capacity = initialCapacity;
+            }
+        }
+
+        /// <summary>
+        /// Returns the count of elements in the list.
+        /// </summary>
+        public int Count
+        {
+            get { return _count; }
+        }
+
+        /// <summary>
+        /// Copy to Array
+        /// </summary>
+        /// <returns>Array copy</returns>
+        public T[] UnderlyingArray => ToArray();
+
+        /// <summary>
+        /// Returns the last element on the list and removes it from it.
+        /// </summary>
+        /// <returns>The last element that was on the list.</returns>
+        public T Pop()
+        {
+            if (_count == 0)
+            {
+                throw new InvalidOperationException("Attempting to remove an element from empty collection.");
+            }
+
+            var oldSegmentIndex = --_count >> _segmentShift;
+            var result = _items[oldSegmentIndex][_count & _offsetMask];
+
+            var newSegmentIndex = (_count - 1) >> _segmentShift;
+
+            if (newSegmentIndex != oldSegmentIndex)
+            {
+                _items[oldSegmentIndex] = null;
```

Comment:
pity to lose the array to GC.  should it be pooled?

Reply:
It's likely gen 0 by the time it's collected. Pooling may be more expensive with write barriers, plus we don't need to worry about LOH like before.